### DISCUSSION TITLE
Sql Alchemy 2.0 Row Type

### DISCUSSION THREAD

**Discussion by martinuncountable** [sqlalchemy/sqlalchemy#10487] [Usage Questions] [answered]
The documentation states that from 1.4 the `Row` object acts like a named tuple. However, the types do not reflect that.

Reviewing some of the issues and discussions, the new `.t` field on `Row` is the only way to get some typing from the `Row` object. Ideally in this example `row.id` would resolve to an `int`:

```
query = session.query(User.id, User.name)
for row in query:
    reveal_type(row)
    reveal_type(row.id)

    # Only way to get some sort of types:
    reveal_type(row.t)

--- mypy gives ---
example.py:27: note: Revealed type is "sqlalchemy.engine.row.Row[tuple[builtins.int, builtins.str]]"
example.py:28: note: Revealed type is "Any"
example.py:31: note: Revealed type is "tuple[builtins.int, builtins.str]"
```

Having an `Any` opens us up to many potential bugs we should be catching during CI.

What do you recommend to do for our use case? Is there active development for resolving Row with a NamedTuple/Dataclass like this?

**Comment by CaselIT** [comment, marked as answer]
Hi,

use `select` instead of query, so you can do `session.execute(a_select_here).tuples()`.
As far as the sqlalchemy developer are aware there isn't currently a way of making `row.id` type correctly, since it's a dynamic variable. That is with the current python typing support, but I doubt it will ever be possible.
The best option would be to have `row[0]` type correctly but that again is not possible with the current typing features of python (again AFAIK)

We are open to improvements on this part btw, so feel free to look into making `row[0]` work

**Comment by zzzeek** [comment]
dupe of #10496
